### ISSUE TITLE
cuda和cpu上精度误差较大

### ISSUE THREAD

**Issue by GG-yuki** [Oneflow-Inc/oneflow#8914]
## Summary

目标是解决一个高精度的模拟计算任务，代码部分的工作包括：1.oneflow重写训练代码；2.oneflow实现LBFGS优化算法，对标torch.optim中的lbfgs优化算法。训练部分代码可参照torch版本https://github.com/jayroxis/PINNs/blob/cee3f2bd20b5751deb3865fbe80cb4794a157933/Burgers%20Equation/Burgers.ipynb。
LBFGS优化算法可参照oneflow版本https://github.com/Oneflow-Inc/oneflow/pull/8750/commits/0243653d1ba2252764dd2cfa9fa883cd1219da55。训练任务较简单，主要问题应该是LBFGS优化算法这块。

目前出现问题是，cpu版本的oneflow目前已调试到精度为99%，cuda版本上则为50%左右（同样的checkpoint）。在cuda上计算时小数点四位之后和cpu上不对齐，LBFGS中会经由不同的if-else分支放大，几千轮训练后最终滚雪球累加到小数点后一位甚至个位数的loss（cpu上loss大致小数点后4位）。

用了官方上手的MNIST的代码也做了同样的测试，5轮测试后，精度差大概在小数点后5位左右，在第一轮grad后即开始出现误差。官方代码为Linear网络，改成卷积网络后也有类似问题，但linear上误差更明显。

此外，torch上也有类似问题，在MNIST上同样的算子在cuda中和cpu上有些微精度损失。在模拟计算任务中，cpu上accuracy为99%，cuda上精度为70%左右。

## Code to reproduce bug

Please post a minimal example to repro the bug. GitHub Gist or repo is highly recommended.

## System Information

- What is your OneFlow installation (pip, source, dockerhub):
- OS:
- OneFlow version (run `python3 -m oneflow --doctor`):
- Python version:
- CUDA driver version:
- GPU models:
- Other info:

**Comment by GG-yuki** [comment]
新建了一个分支，在6月初的master下做了点改动，链接：https://github.com/Oneflow-Inc/oneflow/tree/Lbfgs

[code.zip](https://github.com/Oneflow-Inc/oneflow/files/9348207/code.zip)

训练代码及优化算法代码如附件所示。解压后代码可直接运行pinns_oneflow.py。开头有定义cuda or cpu device。结果会打印出来并以图片形式展现，可以主要关注第二幅图片，即方程解与神经网络拟合函数解的重合程度。

**Comment by GG-yuki** [comment]
tanh_grad报错。具体信息如下：
oneflow._oneflow_internal.exception.Exception: Check failed: (IsClassRegistered<std::string, OpExprGradFunctionIf>(proto().op_type_name())) The gradient function for op tanh_grad is not found. Please check whether it has been implemented and registered correctly.
下边是个报错的小demo。

[test.zip](https://github.com/Oneflow-Inc/oneflow/files/9357621/test.zip)
